janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/strategy/AdjacentVertexHasUniquePropertyOptimizerStrategy.java

```diff
@@ -106,15 +106,13 @@ protected boolean isValidStep(HasStep<?> step) {
             .collect(Collectors.toList());
         final MultiCondition<JanusGraphElement> conditions = QueryUtil.constraints2QNF(tx, constraints);
 
-        // collect all matching indexes
-        final Set<IndexType> availableIndexes = IndexSelectionUtil.getMatchingIndexes(conditions);
-
-        // select only valid unique indexes
-        return availableIndexes.stream()
-            .filter(IndexType::isCompositeIndex)
-            .map(index -> (CompositeIndexType) index)
-            .filter(cIndex -> cIndex.getCardinality() == Cardinality.SINGLE)
-            .anyMatch(cIndex -> IndexSelectionUtil.isIndexSatisfiedByGivenKeys(cIndex, givenKeys));
+        // collect all matching unique indexes
+        final Set<IndexType> availableIndexes = IndexSelectionUtil.getMatchingIndexes(conditions,
+            indexType -> indexType.isCompositeIndex()
+                && ((CompositeIndexType) indexType).getCardinality() == Cardinality.SINGLE
+                && IndexSelectionUtil.isIndexSatisfiedByGivenKeys(indexType, givenKeys));
+
+        return !availableIndexes.isEmpty();
```

Comment:
(nitpick)
We don't necessary need to collect all available indices here. All we need is just answer if there is any index. I.e. we don't actually need to always process all `conditions`. Instead it's enough to process up to the first index which matches out filter. After that we can immediately return `true`.

Reply:
Yes that is correct. That was potentially even better in the previous version using Java streams and `anyMatch()`. In that implementation the processing can stop after the first match has been found.

Reply:
Well, in the previous version you were still creating the full Set. So, no, it wasn't better in the previous version. It is better now actually. But again, I just pointed it because we may improve it even more (we don't need to create any `Set` here. We can just iterate through `conditions` till we find the first condition which meets our criteria)

Reply:
Previously you had:
```
// This takes N time
final Set<IndexType> availableIndexes = IndexSelectionUtil.getMatchingIndexes(conditions);

// This takes N time in the worst scenario
return availableIndexes.stream().filter(IndexType::isCompositeIndex).map(index -> (CompositeIndexType) index).filter(cIndex -> cIndex.getCardinality() == Cardinality.SINGLE).anyMatch(cIndex -> IndexSelectionUtil.isIndexSatisfiedByGivenKeys(cIndex, givenKeys)); 
```
So, previously you had N iterations in the best scenario and N*2 iterations in the worst scenario.

Now you had N iterations in both best and worst scenario, which is better then the previous logic.
What I'm telling is that we can improve best case scenario here. We can reduce the amount of iteration in the best scenario to 1 instead of N. That say, the worst scenario will take N iterations still - which is fine.

Again, I'm not telling that we need to do it in this PR. I'm fine doing it in another PR or changing it in this PR.

Reply:
The full set was created in the previous version but note every entry of the set was necessarily accessed to check the conditions. In this implementation now, every check is executed before adding the index candidates to the set. Thus, the set is more selective, but more checks are executed overall.

Reply:
That's true. As I said, we don't need any `Set` here. We can simply implement loop iteration over `conditions` till the first matched condition (something like `anyMatch` in streams). Thus, our time and space will be more efficient.
I.e. we don't need to use `getMatchingIndexes` here at all

Reply:
That being said, you are obviously right that in this case, the set can be omitted entirely. My initial idea was to externalize the set creation to a method which can then be used for both use cases. It's a matter of design whether or not we want to duplicate the logic of this method or if we sacrifice that bit of efficiency for maintainability. I have a slight preference for maintainability and more general purpose methods but that's not set in stone.

Reply:
@rngcntr Makes sense. In such case, I think it makes sense to revert changes in this method back to not execute `IndexSelectionUtil.isIndexSatisfiedByGivenKeys(indexType, givenKeys)` for each condition

Reply:
I'm just working on a draft for a method like `existsMatchingIndex` as you suggested earlier, just do see how much duplicate code it would cause. Stand by...

Reply:
Alright, first draft is pushed. I further externalized some of the logic do avoid duplicate code.